ICE: compiler/rustc_middle/src/ty/generics.rs:188:18: expected type parameter, but found another generic parameter

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
extern "platform-intrinsic" {
    fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.54.0-nightly (c1e8f3a58 2021-05-30)
binary: rustc
commit-hash: c1e8f3a5857c3c8d095fd212acc6ed442f4cc58b
commit-date: 2021-05-30
host: x86_64-unknown-linux-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

### Error output

```
error[E0658]: platform intrinsics are experimental and possibly buggy
 --> mutant.rs:1:8
  |
1 | extern "platform-intrinsic" {
  |        ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #27731 <https://github.com/rust-lang/rust/issues/27731> for more information
  = help: add `#![feature(platform_intrinsics)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `mutant`
 --> mutant.rs:1:1
  |
1 | / extern "platform-intrinsic" {
2 | |     fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
3 | | }
  | |_^ consider adding a `main` function to `mutant.rs`

error: internal compiler error: compiler/rustc_middle/src/ty/generics.rs:188:18: expected type parameter, but found another generic parameter

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:1005:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (c1e8f3a58 2021-05-30) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_mod_item_types] checking item types in top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error[E0658]: platform intrinsics are experimental and possibly buggy
 --> mutant.rs:1:8
  |
1 | extern "platform-intrinsic" {
  |        ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #27731 <https://github.com/rust-lang/rust/issues/27731> for more information
  = help: add `#![feature(platform_intrinsics)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `mutant`
 --> mutant.rs:1:1
  |
1 | / extern "platform-intrinsic" {
2 | |     fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
3 | | }
  | |_^ consider adding a `main` function to `mutant.rs`

error: internal compiler error: compiler/rustc_middle/src/ty/generics.rs:188:18: expected type parameter, but found another generic parameter

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:1005:9
stack backtrace:
   0: std::panicking::begin_panic
   1: std::panic::panic_any
   2: rustc_errors::HandlerInner::bug
   3: rustc_errors::Handler::bug
   4: rustc_middle::ty::context::tls::with_opt
   5: rustc_middle::util::bug::opt_span_bug_fmt
   6: rustc_middle::util::bug::bug_fmt
   7: rustc_middle::ty::generics::Generics::type_param
   8: rustc_middle::ty::error::<impl rustc_middle::ty::context::TyCtxt>::note_and_explain_type_err
   9: rustc_infer::infer::error_reporting::<impl rustc_infer::infer::InferCtxt>::note_type_err
  10: rustc_infer::infer::error_reporting::<impl rustc_infer::infer::InferCtxt>::report_and_explain_type_error
  11: rustc_infer::infer::InferCtxt::report_mismatched_types
  12: rustc_infer::infer::InferCtxtBuilder::enter
  13: rustc_typeck::check::intrinsic::equate_intrinsic_type
  14: rustc_typeck::check::intrinsic::check_platform_intrinsic_type
  15: rustc_typeck::check::check::check_item_type
  16: rustc_middle::hir::map::Map::visit_item_likes_in_module
  17: rustc_typeck::check::check::check_mod_item_types
  18: rustc_query_system::query::plumbing::get_query_impl
  19: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::check_mod_item_types
  20: rustc_session::utils::<impl rustc_session::session::Session>::time
  21: rustc_typeck::check_crate
  22: rustc_interface::passes::analysis
  23: rustc_query_system::query::plumbing::get_query_impl
  24: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  25: rustc_interface::passes::QueryContext::enter
  26: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  27: rustc_span::with_source_map
  28: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (c1e8f3a58 2021-05-30) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_mod_item_types] checking item types in top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.
```

</p>
</details>

--- a/compiler/rustc_error_codes/src/error_codes/E0094.md
+++ b/compiler/rustc_error_codes/src/error_codes/E0094.md
@@ -1,4 +1,5 @@
-An invalid number of type parameters was given to an intrinsic function.
+An invalid number of generic type, lifetime, or const parameters was
+given to an intrinsic function.
 
 Erroneous code example:
 

--- a/compiler/rustc_error_codes/src/error_codes/E0094.md
+++ b/compiler/rustc_error_codes/src/error_codes/E0094.md
@@ -1,5 +1,4 @@
-An invalid number of generic type, lifetime, or const parameters was
-given to an intrinsic function.
+An invalid number of generic parameters was passed to an intrinsic function.
 
 Erroneous code example:
 

--- a/compiler/rustc_typeck/src/check/intrinsic.rs
+++ b/compiler/rustc_typeck/src/check/intrinsic.rs
@@ -3,11 +3,11 @@
 
 use crate::errors::{
     SimdShuffleMissingLength, UnrecognizedAtomicOperation, UnrecognizedIntrinsicFunction,
-    WrongNumberOfTypeArgumentsToInstrinsic,
+    WrongNumberOfGenericArgumentsToInstrinsic,
 };
 use crate::require_same_types;
 
-use rustc_errors::struct_span_err;
+use rustc_errors::{pluralize, struct_span_err};
 use rustc_hir as hir;
 use rustc_middle::traits::{ObligationCause, ObligationCauseCode};
 use rustc_middle::ty::subst::Subst;
@@ -21,36 +21,68 @@ fn equate_intrinsic_type<'tcx>(
     tcx: TyCtxt<'tcx>,
     it: &hir::ForeignItem<'_>,
     n_tps: usize,
+    n_lts: usize,
     sig: ty::PolyFnSig<'tcx>,
 ) {
-    match it.kind {
-        hir::ForeignItemKind::Fn(..) => {}
+    let (gen_lts, gen_tys, gen_cns, span) = match &it.kind {
+        hir::ForeignItemKind::Fn(.., generics) => {
+            let mut gen_lts = 0;
+            let mut gen_tys = 0;
+            let mut gen_cns = 0;
+
+            for param in generics.params {
+                match param.kind {
+                    hir::GenericParamKind::Lifetime { .. } => {
+                        gen_lts += 1;
+                    }
+                    hir::GenericParamKind::Type { .. } => {
+                        gen_tys += 1;
+                    }
+                    hir::GenericParamKind::Const { .. } => {
+                        gen_cns += 1;
+                    }
+                }
+            }
+
+            (gen_lts, gen_tys, gen_cns, generics.span)
+        }
         _ => {
             struct_span_err!(tcx.sess, it.span, E0622, "intrinsic must be a function")
                 .span_label(it.span, "expected a function")
                 .emit();
             return;
         }
-    }
+    };
 
-    let i_n_tps = tcx.generics_of(it.def_id).own_counts().types;
-    if i_n_tps != n_tps {
-        let span = match it.kind {
-            hir::ForeignItemKind::Fn(_, _, ref generics) => generics.span,
-            _ => bug!(),
-        };
-
-        tcx.sess.emit_err(WrongNumberOfTypeArgumentsToInstrinsic {
+    if gen_lts != n_lts {
+        tcx.sess.emit_err(WrongNumberOfGenericArgumentsToInstrinsic {
             span,
-            found: i_n_tps,
-            expected: n_tps,
+            found: gen_lts,
+            expected: n_lts,
+            expected_pluralize: pluralize!(n_lts),
+            descr: "lifetime",
         });
-        return;
+    } else if gen_tys != n_tps {
+        tcx.sess.emit_err(WrongNumberOfGenericArgumentsToInstrinsic {
+            span,
+            found: gen_tys,
+            expected: n_tps,
+            expected_pluralize: pluralize!(n_tps),
+            descr: "type",
+        });
+    } else if gen_cns != 0 {
+        tcx.sess.emit_err(WrongNumberOfGenericArgumentsToInstrinsic {
+            span,
+            found: gen_cns,
+            expected: 0,
+            expected_pluralize: pluralize!(0),
+            descr: "const",
+        });
+    } else {
+        let fty = tcx.mk_fn_ptr(sig);
+        let cause = ObligationCause::new(it.span, it.hir_id(), ObligationCauseCode::IntrinsicType);
+        require_same_types(tcx, &cause, tcx.mk_fn_ptr(tcx.fn_sig(it.def_id)), fty);
     }
-
-    let fty = tcx.mk_fn_ptr(sig);
-    let cause = ObligationCause::new(it.span, it.hir_id(), ObligationCauseCode::IntrinsicType);
-    require_same_types(tcx, &cause, tcx.mk_fn_ptr(tcx.fn_sig(it.def_id)), fty);
 }
 
 /// Returns `true` if the given intrinsic is unsafe to call or not.
@@ -121,7 +153,7 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>) {
         })
     };
 
-    let (n_tps, inputs, output, unsafety) = if name_str.starts_with("atomic_") {
+    let (n_tps, n_lts, inputs, output, unsafety) = if name_str.starts_with("atomic_") {
         let split: Vec<&str> = name_str.split('_').collect();
         assert!(split.len() >= 2, "Atomic intrinsic in an incorrect format");
 
@@ -143,7 +175,7 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>) {
                 return;
             }
         };
-        (n_tps, inputs, output, hir::Unsafety::Unsafe)
+        (n_tps, 0, inputs, output, hir::Unsafety::Unsafe)
     } else {
         let unsafety = intrinsic_operation_unsafety(intrinsic_name);
         let (n_tps, inputs, output) = match intrinsic_name {
@@ -372,11 +404,17 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>) {
                 return;
             }
         };
-        (n_tps, inputs, output, unsafety)
+        (
+            n_tps,
+            if matches!(intrinsic_name, sym::va_copy) { 1 } else { 0 },
+            inputs,
+            output,
+            unsafety,
+        )
     };
     let sig = tcx.mk_fn_sig(inputs.into_iter(), output, false, unsafety, Abi::RustIntrinsic);
     let sig = ty::Binder::bind_with_vars(sig, bound_vars);
-    equate_intrinsic_type(tcx, it, n_tps, sig)
+    equate_intrinsic_type(tcx, it, n_tps, n_lts, sig)
 }
 
 /// Type-check `extern "platform-intrinsic" { ... }` functions.
@@ -472,5 +510,5 @@ pub fn check_platform_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>)
         Abi::PlatformIntrinsic,
     );
     let sig = ty::Binder::dummy(sig);
-    equate_intrinsic_type(tcx, it, n_tps, sig)
+    equate_intrinsic_type(tcx, it, n_tps, 0, sig)
 }

--- a/compiler/rustc_typeck/src/check/intrinsic.rs
+++ b/compiler/rustc_typeck/src/check/intrinsic.rs
@@ -3,7 +3,7 @@
 
 use crate::errors::{
     SimdShuffleMissingLength, UnrecognizedAtomicOperation, UnrecognizedIntrinsicFunction,
-    WrongNumberOfGenericArgumentsToInstrinsic,
+    WrongNumberOfGenericArgumentsToIntrinsic,
 };
 use crate::require_same_types;
 
@@ -24,27 +24,10 @@ fn equate_intrinsic_type<'tcx>(
     n_lts: usize,
     sig: ty::PolyFnSig<'tcx>,
 ) {
-    let (gen_lts, gen_tys, gen_cns, span) = match &it.kind {
+    let (own_counts, span) = match &it.kind {
         hir::ForeignItemKind::Fn(.., generics) => {
-            let mut gen_lts = 0;
-            let mut gen_tys = 0;
-            let mut gen_cns = 0;
-
-            for param in generics.params {
-                match param.kind {
-                    hir::GenericParamKind::Lifetime { .. } => {
-                        gen_lts += 1;
-                    }
-                    hir::GenericParamKind::Type { .. } => {
-                        gen_tys += 1;
-                    }
-                    hir::GenericParamKind::Const { .. } => {
-                        gen_cns += 1;
-                    }
-                }
-            }
-
-            (gen_lts, gen_tys, gen_cns, generics.span)
+            let own_counts = tcx.generics_of(it.def_id.to_def_id()).own_counts();
+            (own_counts, generics.span)
         }
         _ => {
             struct_span_err!(tcx.sess, it.span, E0622, "intrinsic must be a function")
@@ -54,31 +37,25 @@ fn equate_intrinsic_type<'tcx>(
         }
     };
 
-    if gen_lts != n_lts {
-        tcx.sess.emit_err(WrongNumberOfGenericArgumentsToInstrinsic {
-            span,
-            found: gen_lts,
-            expected: n_lts,
-            expected_pluralize: pluralize!(n_lts),
-            descr: "lifetime",
-        });
-    } else if gen_tys != n_tps {
-        tcx.sess.emit_err(WrongNumberOfGenericArgumentsToInstrinsic {
-            span,
-            found: gen_tys,
-            expected: n_tps,
-            expected_pluralize: pluralize!(n_tps),
-            descr: "type",
-        });
-    } else if gen_cns != 0 {
-        tcx.sess.emit_err(WrongNumberOfGenericArgumentsToInstrinsic {
-            span,
-            found: gen_cns,
-            expected: 0,
-            expected_pluralize: pluralize!(0),
-            descr: "const",
-        });
-    } else {
+    let gen_count_ok = |found: usize, expected: usize, descr: &str| -> bool {
+        if found != expected {
+            tcx.sess.emit_err(WrongNumberOfGenericArgumentsToIntrinsic {
+                span,
+                found,
+                expected,
+                expected_pluralize: pluralize!(expected),
+                descr,
+            });
+            false
+        } else {
+            true
+        }
+    };
+
+    if gen_count_ok(own_counts.lifetimes, n_lts, "lifetime")
+        && gen_count_ok(own_counts.types, n_tps, "type")
+        && gen_count_ok(own_counts.consts, 0, "const")
+    {
         let fty = tcx.mk_fn_ptr(sig);
         let cause = ObligationCause::new(it.span, it.hir_id(), ObligationCauseCode::IntrinsicType);
         require_same_types(tcx, &cause, tcx.mk_fn_ptr(tcx.fn_sig(it.def_id)), fty);
@@ -404,13 +381,7 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>) {
                 return;
             }
         };
-        (
-            n_tps,
-            if matches!(intrinsic_name, sym::va_copy) { 1 } else { 0 },
-            inputs,
-            output,
-            unsafety,
-        )
+        (n_tps, 0, inputs, output, unsafety)
     };
     let sig = tcx.mk_fn_sig(inputs.into_iter(), output, false, unsafety, Abi::RustIntrinsic);
     let sig = ty::Binder::bind_with_vars(sig, bound_vars);

--- a/compiler/rustc_typeck/src/errors.rs
+++ b/compiler/rustc_typeck/src/errors.rs
@@ -24,13 +24,15 @@ pub struct UnrecognizedAtomicOperation<'a> {
 
 #[derive(SessionDiagnostic)]
 #[error = "E0094"]
-pub struct WrongNumberOfTypeArgumentsToInstrinsic {
-    #[message = "intrinsic has wrong number of type \
+pub struct WrongNumberOfGenericArgumentsToInstrinsic<'a> {
+    #[message = "intrinsic has wrong number of {descr} \
                          parameters: found {found}, expected {expected}"]
-    #[label = "expected {expected} type parameter"]
+    #[label = "expected {expected} {descr} parameter{expected_pluralize}"]
     pub span: Span,
     pub found: usize,
     pub expected: usize,
+    pub expected_pluralize: &'a str,
+    pub descr: &'a str,
 }
 
 #[derive(SessionDiagnostic)]

--- a/compiler/rustc_typeck/src/errors.rs
+++ b/compiler/rustc_typeck/src/errors.rs
@@ -24,7 +24,7 @@ pub struct UnrecognizedAtomicOperation<'a> {
 
 #[derive(SessionDiagnostic)]
 #[error = "E0094"]
-pub struct WrongNumberOfGenericArgumentsToInstrinsic<'a> {
+pub struct WrongNumberOfGenericArgumentsToIntrinsic<'a> {
     #[message = "intrinsic has wrong number of {descr} \
                          parameters: found {found}, expected {expected}"]
     #[label = "expected {expected} {descr} parameter{expected_pluralize}"]

--- a/src/test/ui/simd-intrinsic/issue-85855.rs
+++ b/src/test/ui/simd-intrinsic/issue-85855.rs
@@ -1,0 +1,20 @@
+// Check that appropriate errors are reported if an intrinsic is defined
+// with the wrong number of generic lifetime/type/const parameters, and
+// that no ICE occurs in these cases.
+
+#![feature(platform_intrinsics)]
+#![crate_type="lib"]
+
+extern "platform-intrinsic" {
+    fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
+    //~^ ERROR: intrinsic has wrong number of lifetime parameters
+
+    fn simd_add<'a, T>(x: T, y: T);
+    //~^ ERROR: intrinsic has wrong number of lifetime parameters
+
+    fn simd_sub<T, U>(x: T, y: U);
+    //~^ ERROR: intrinsic has wrong number of type parameters
+
+    fn simd_mul<T, const N: usize>(x: T, y: T);
+    //~^ ERROR: intrinsic has wrong number of const parameters
+}

--- a/src/test/ui/simd-intrinsic/issue-85855.rs
+++ b/src/test/ui/simd-intrinsic/issue-85855.rs
@@ -9,8 +9,7 @@ extern "platform-intrinsic" {
     fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
     //~^ ERROR: intrinsic has wrong number of lifetime parameters
 
-    fn simd_add<'a, T>(x: T, y: T);
-    //~^ ERROR: intrinsic has wrong number of lifetime parameters
+    fn simd_add<'a, T>(x: T, y: T) -> T;
 
     fn simd_sub<T, U>(x: T, y: U);
     //~^ ERROR: intrinsic has wrong number of type parameters

--- a/src/test/ui/simd-intrinsic/issue-85855.stderr
+++ b/src/test/ui/simd-intrinsic/issue-85855.stderr
@@ -1,0 +1,27 @@
+error[E0094]: intrinsic has wrong number of lifetime parameters: found 1, expected 0
+  --> $DIR/issue-85855.rs:9:27
+   |
+LL |     fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
+   |                           ^^^^^^^^^^^ expected 0 lifetime parameters
+
+error[E0094]: intrinsic has wrong number of lifetime parameters: found 1, expected 0
+  --> $DIR/issue-85855.rs:12:16
+   |
+LL |     fn simd_add<'a, T>(x: T, y: T);
+   |                ^^^^^^^ expected 0 lifetime parameters
+
+error[E0094]: intrinsic has wrong number of type parameters: found 2, expected 1
+  --> $DIR/issue-85855.rs:15:16
+   |
+LL |     fn simd_sub<T, U>(x: T, y: U);
+   |                ^^^^^^ expected 1 type parameter
+
+error[E0094]: intrinsic has wrong number of const parameters: found 1, expected 0
+  --> $DIR/issue-85855.rs:18:16
+   |
+LL |     fn simd_mul<T, const N: usize>(x: T, y: T);
+   |                ^^^^^^^^^^^^^^^^^^^ expected 0 const parameters
+
+error: aborting due to 4 previous errors
+
+For more information about this error, try `rustc --explain E0094`.

--- a/src/test/ui/simd-intrinsic/issue-85855.stderr
+++ b/src/test/ui/simd-intrinsic/issue-85855.stderr
@@ -4,24 +4,18 @@ error[E0094]: intrinsic has wrong number of lifetime parameters: found 1, expect
 LL |     fn simd_saturating_add<'a, T: 'a>(x: T, y: T);
    |                           ^^^^^^^^^^^ expected 0 lifetime parameters
 
-error[E0094]: intrinsic has wrong number of lifetime parameters: found 1, expected 0
-  --> $DIR/issue-85855.rs:12:16
-   |
-LL |     fn simd_add<'a, T>(x: T, y: T);
-   |                ^^^^^^^ expected 0 lifetime parameters
-
 error[E0094]: intrinsic has wrong number of type parameters: found 2, expected 1
-  --> $DIR/issue-85855.rs:15:16
+  --> $DIR/issue-85855.rs:14:16
    |
 LL |     fn simd_sub<T, U>(x: T, y: U);
    |                ^^^^^^ expected 1 type parameter
 
 error[E0094]: intrinsic has wrong number of const parameters: found 1, expected 0
-  --> $DIR/issue-85855.rs:18:16
+  --> $DIR/issue-85855.rs:17:16
    |
 LL |     fn simd_mul<T, const N: usize>(x: T, y: T);
    |                ^^^^^^^^^^^^^^^^^^^ expected 0 const parameters
 
-error: aborting due to 4 previous errors
+error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0094`.

